Mathlib/RingTheory/Length.lean

```diff
@@ -121,3 +156,96 @@ lemma Module.length_eq_add_of_exact :
   · have := mt (IsFiniteLength.of_surjective · hg) hP
     rw [← Module.length_ne_top_iff, ne_eq, not_not] at hP this
     rw [hP, this, add_top]
+
+include hf in
+lemma Module.length_le_of_injective : Module.length R N ≤ Module.length R M := by
+  rw [Module.length_eq_add_of_exact f (LinearMap.range f).mkQ hf
+    (Submodule.mkQ_surjective _) (LinearMap.exact_map_mkQ_range f)]
+  exact le_self_add
+
+include hg in
+lemma Module.length_le_of_surjective : Module.length R P ≤ Module.length R M := by
+  rw [Module.length_eq_add_of_exact (LinearMap.ker g).subtype g (Submodule.subtype_injective _) hg
+    (LinearMap.exact_subtype_ker_map g)]
+  exact le_add_self
+
+variable (R M N) in
+@[simp]
+lemma Module.length_prod :
+    Module.length R (M × N) = Module.length R M + Module.length R N :=
+  Module.length_eq_add_of_exact _ _ LinearMap.inl_injective LinearMap.snd_surjective .inl_snd
+
+variable (R) in
+@[simp]
+lemma Module.length_pi_of_fintype {ι : Type*} [Fintype ι]
+    (M : ι → Type*) [∀ i, AddCommGroup (M i)] [∀ i, Module R (M i)] :
+    Module.length R (Π i, M i) = ∑ i, Module.length R (M i) := by
+  refine Fintype.induction_empty_option (P := fun ι (_ : Fintype ι) ↦
+    ∀ (M : ι → Type _) [∀ i, AddCommGroup (M i)] [∀ i, Module R (M i)],
+      Module.length R (Π i, M i) = ∑ i, Module.length R (M i)) ?_ ?_ ?_ ι M
+  · intro α β _ e IH M _ _
+    let _ : Fintype α := .ofEquiv β e.symm
+    rw [← (LinearEquiv.piCongrLeft R M e).length_eq, IH, e.sum_comp (length R <| M ·)]
+  · intro M _ _
+    simp [Module.length_eq_zero]
+  · intro ι _ IH M _ _
+    rw [(LinearEquiv.piOptionEquivProd _).length_eq, Module.length_prod, IH, add_comm,
+      Fintype.sum_option, add_comm]
+
+@[simp]
+lemma Module.length_finsupp {ι : Type*} :
+    Module.length R (ι →₀ M) = ENat.card ι * Module.length R M := by
+  cases finite_or_infinite ι
+  · cases nonempty_fintype ι
+    simp [(Finsupp.linearEquivFunOnFinite R M ι).length_eq]
+  nontriviality M
+  rw [ENat.card_eq_top_of_infinite, ENat.top_mul length_pos.ne', ENat.eq_top_iff_forall_ge]
+  intro m
+  obtain ⟨s, hs⟩ := Infinite.exists_subset_card_eq ι m
+  have : length R (s →₀ M) = ↑m * length R M := by
+    simp [(Finsupp.linearEquivFunOnFinite R M _).length_eq, hs]
+  refine le_trans ?_ (Module.length_le_of_injective (Finsupp.lmapDomain M R ((↑) : s → ι))
+    (Finsupp.mapDomain_injective Subtype.val_injective))
+  rw [this]
+  exact ENat.self_le_mul_right _ length_pos.ne'
+
+@[simp]
+lemma Module.length_pi {ι : Type*} [Finite ι] :
+    Module.length R (ι → M) = ENat.card ι * Module.length R M := by
```

Comment:
This is still true if `ι` is infinite because [Finsupp.lcoeFun](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Finsupp/Pi.html#Finsupp.lcoeFun) is injective.

Reply:
Oh, you already proved it. You just need to remove the `[Finite ι]`.